### DISCUSSION TITLE
테스트 코드에서 리플렉션을 이용한 extracting 메서드를 어떻게 생각하시나요?

### DISCUSSION THREAD

**Discussion by gyungchan-jo** [2024-woowacourse-study/level-interview#104] [Test]
.

**Comment by jimi567** [comment]
extracting으로 테스트를 작성하게 되면, 불필요한 getter나 테스트를 위한 프로덕션 코드를 최소화할 수 있다고 생각합니다. 또, 리플랙션을 사용하여 메서드를 추출하면, 해당 메서드를 여러 테스트 케이스에서 재사용할 수 있고, 이는 중복 코드를 줄여주기 때문에 테스트 작성에 드는 리소스가 적어집니다.

하지만, 신뢰성이 매우 낮다고 생각합니다. 단순히 변수명 변경에도 , 테스트가 실패할 가능성이 있습니다.
또 리플랙션은 실행 시간에 클래스 정보를 동적으로 조작하기 떄문에 성능측면에서 단점이 명확하다고 생각합니다. 
리플랙션을 사용한 테스트 코드는 가독성 측면에서도 단점이 크게 느껴져서, 
잘 사용하지는 않습니다.

**Comment by nayonsoso** [comment]
private 메서드를 테스트하기 위해 extracting을 사용하는 경우가 종종 있습니다.
하지만 저는 이런 테스트 코드는 내부 구현에 의존하기 때문에 유지보수가 어려워진다고 생각합니다. 
또 캡슐화를 위배한다는 점도 문제가 될 것 같습니다.

테스트 코드를 작성하는 본질은 '기능'을 테스트하기 위함입니다.
그리고 이 '기능'이라고 하는 것은 우리가 호출해서 사용할 수 있는 public 메서드라고 생각합니다.
public 메서드를 테스트한다면, 그 내부에서 호출되는 pricate 메서드들에 대해서는 자연스럽게 검증이 된다고 생각합니다.
따라서 public 메서드의 테스트로만 충분하는게 제 생각입니다.

**Comment by GIVEN53** [comment]
테스트를 위한 메서드 임시 방편
필드명 하드 코딩 -> 테스트 깨지기 쉬움
무분별 사용 x
